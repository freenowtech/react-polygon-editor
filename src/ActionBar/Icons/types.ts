import { SVGProps } from 'react';

export interface IconProps extends SVGProps<SVGSVGElement> {
    iconColor?: string;
}
