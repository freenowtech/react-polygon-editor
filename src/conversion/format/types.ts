import { Coordinate } from 'types';

export enum FormatType {
    GEOJSON = 'geojson',
    LATLNG = 'latlng',
    JTS = 'jts',
}

export interface Format {
    name: FormatType;
    displayName: string;
    description?: string;
    serialize: (coordinates: Coordinate[]) => string;
    deserialize: (raw: string) => Coordinate[];
    validate: (value: string) => boolean;
}
