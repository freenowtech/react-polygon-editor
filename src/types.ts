import { ActionCreatorsMapObject } from 'redux';

export type Coordinate = {
    latitude: number;
    longitude: number;
};

export interface Action<T extends string> {
    type: T;
}

export interface ActionWithPayload<T extends string, P> extends Action<T> {
    payload: P;
}

export type ActionsUnion<A extends ActionCreatorsMapObject> = ReturnType<A[keyof A]>;
