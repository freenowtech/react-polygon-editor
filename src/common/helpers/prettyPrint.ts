// tslint:disable-next-line: no-any
export const prettyPrint = (value: any) => JSON.stringify(value, null, 2);
