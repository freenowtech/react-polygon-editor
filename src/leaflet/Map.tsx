import styled from 'styled-components';

import { MapContainer as LeafletMap } from 'react-leaflet';

export const Map = styled(LeafletMap)`
    height: 100%;
    width: 100%;
    z-index: 0;

    cursor: ${({ drawCursor }: { drawCursor: boolean }) => (drawCursor ? 'crosshair !important' : 'auto')};
`;

export const Container = styled.div`
    position: relative;
    height: 100%;
    width: 100%;
`;
