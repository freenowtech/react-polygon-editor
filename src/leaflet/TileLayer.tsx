import React from 'react';
import { TileLayer as LeafletTileLayer } from 'react-leaflet';

import { LEAFLET } from '../constants';

export const TileLayer = () => (
    <LeafletTileLayer
        attribution={'&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'}
        url={LEAFLET.TILE_SERVER_URL}
    />
);
