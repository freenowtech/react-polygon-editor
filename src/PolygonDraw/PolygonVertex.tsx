import 'leaflet-path-drag';
import React from 'react';
import { CircleMarker, LatLng } from 'leaflet';
import { CircleMarker as ReactLeafletCircleMarker } from 'react-leaflet';

import { Coordinate } from 'types';
import { createLeafletLatLngFromCoordinate } from '../helpers';
import { MAP } from '../constants';

type DragEvent = { target: { getLatLng: () => LatLng } };

export interface Props {
    coordinate: Coordinate;
    isSelected: boolean;
    index: number;
    onClick: (index: number) => void;
    onDragStart: (latlng: LatLng, index: number) => void;
    onDrag: (latlng: LatLng) => void;
    onDragEnd: () => void;
}

export interface State {
    isHovered: boolean;
    isDragged: boolean;
    latLng: LatLng;
}

export class PolygonVertex extends React.Component<Props, State> {
    private circleMarkerElement: CircleMarker;
    state = {
        isHovered: false,
        isDragged: false,
        latLng: new LatLng(0, 0),
    };

    static getDerivedStateFromProps = (props: Props, state: State) => {
        return state.isDragged ? state : { ...state, latLng: createLeafletLatLngFromCoordinate(props.coordinate) };
    };

    componentDidMount() {
        if (this.circleMarkerElement) {
            this.circleMarkerElement.on('dragstart', this.handleDragStart);
            this.circleMarkerElement.on('drag', this.handleDrag);
            this.circleMarkerElement.on('dragend', this.handleDragEnd);
        }
    }

    componentWillUnmount() {
        if (this.circleMarkerElement) {
            this.circleMarkerElement.off('dragstart', this.handleDragStart);
            this.circleMarkerElement.off('drag', this.handleDrag);
            this.circleMarkerElement.off('dragend', this.handleDragEnd);
        }
    }

    // tslint:disable-next-line
    setCircleMarkerRef = (ref: any) => {
        if (ref) {
            this.circleMarkerElement = ref;
        }
    };

    handleClick = () => this.props.onClick(this.props.index);
    handleMouseOver = () => this.setState({ isHovered: true });
    handleMouseOut = () => this.setState({ isHovered: false });

    handleDragStart = (event: DragEvent) => {
        this.props.onDragStart(event.target.getLatLng(), this.props.index);
        this.setState({ isDragged: true });
    };

    handleDrag = (event: DragEvent) => {
        if (this.state.isDragged) {
            this.props.onDrag(event.target.getLatLng());
        }
    };

    handleDragEnd = () => {
        this.props.onDragEnd();
        this.setState({ isDragged: false });
    };

    render() {
        const { isSelected } = this.props;
        const { isHovered, latLng } = this.state;
        const selectedOrHovered = isSelected || isHovered;
        const hoveredAndNotSelected = isHovered && !isSelected;

        return (
            <ReactLeafletCircleMarker
                aria-label="Polygon"
                ref={this.setCircleMarkerRef}
                fillColor={MAP.VERTEX_FILL_COLOR}
                fillOpacity={1}
                color={MAP.POLYGON_ACTIVE_COLOR}
                opacity={hoveredAndNotSelected ? 0.6 : 1}
                weight={selectedOrHovered ? 4 : 1}
                radius={selectedOrHovered ? 6 : 4}
                center={latLng}
                eventHandlers={{
                    click: this.handleClick,
                    mouseover: this.handleMouseOver,
                    mouseout: this.handleMouseOut,
                }}
                bubblingMouseEvents={false}
                // @ts-ignore
                draggable
            />
        );
    }
}
