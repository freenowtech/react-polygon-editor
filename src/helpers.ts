import { LatLng, LatLngBounds, LatLngTuple } from 'leaflet';
import isEqual from 'lodash.isequal';

import { Coordinate } from './types';

export const createLeafletLatLngTupleFromCoordinate = (coordinate: Coordinate): LatLngTuple => [
    coordinate.latitude,
    coordinate.longitude,
];

export const createLeafletLatLngBoundsFromCoordinates = (coordinates: Coordinate[]) =>
    new LatLngBounds(coordinates.map(createLeafletLatLngTupleFromCoordinate));

export const createLeafletLatLngFromCoordinate = (coordinate: Coordinate) =>
    new LatLng(coordinate.latitude, coordinate.longitude);

export const createCoordinateFromLeafletLatLng = (latLng: LatLng): Coordinate => ({
    latitude: latLng.lat,
    longitude: latLng.lng,
});

export const addCoordinates = (coordA: Coordinate, coordB: Coordinate): Coordinate => ({
    latitude: coordA.latitude + coordB.latitude,
    longitude: coordA.longitude + coordB.longitude,
});

export const subtractCoordinates = (coordA: Coordinate, coordB: Coordinate): Coordinate => ({
    latitude: coordA.latitude - coordB.latitude,
    longitude: coordA.longitude - coordB.longitude,
});

export const isPolygonClosed = (coordinates: Coordinate[]): boolean =>
    coordinates && coordinates.length > 2 && isEqual(coordinates[0], coordinates[coordinates.length - 1]);

export const isClosingPointsSelected = (coordinates: Coordinate[], selection: Set<number>): boolean =>
    isPolygonClosed(coordinates) && (selection.has(coordinates.length - 1) || selection.has(0));

// https://wrf.ecse.rpi.edu//Research/Short_Notes/pnpoly.html
export const isCoordinateInPolygon = (coordinate: Coordinate, polygon: Coordinate[]): boolean => {
    let inside = false;
    for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
        const intersect =
            polygon[i].longitude > coordinate.longitude !== polygon[j].longitude > coordinate.longitude &&
            coordinate.latitude <
                ((polygon[j].latitude - polygon[i].latitude) * (coordinate.longitude - polygon[i].longitude)) /
                    (polygon[j].longitude - polygon[i].longitude) +
                    polygon[i].latitude;
        if (intersect) {
            inside = !inside;
        }
    }

    return inside;
};

export const movePolygonCoordinates = (
    polygon: Coordinate[],
    selectedCoordinates: Set<number>,
    moveVector: Coordinate
) => {
    const selection = new Set([...selectedCoordinates]);

    if (isClosingPointsSelected(polygon, selection)) {
        selection.add(0);
        selection.add(polygon.length - 1);
    }

    return polygon.map((coord, index) => {
        if (selection.has(index)) {
            return addCoordinates(coord, moveVector);
        } else {
            return coord;
        }
    });
};

export const removeSelectedPoints = (polygonCoordinates: Coordinate[], selectedPoints: Set<number>) => {
    const newPolygonCoordinates = polygonCoordinates.filter((polygonCoordinate, index) => !selectedPoints.has(index));
    const isOldPathClosed =
        polygonCoordinates.length > 1 &&
        isEqual(polygonCoordinates[0], polygonCoordinates[polygonCoordinates.length - 1]);
    const isNewPathClosed =
        newPolygonCoordinates.length > 1 &&
        isEqual(newPolygonCoordinates[0], newPolygonCoordinates[newPolygonCoordinates.length - 1]);

    // Open closed path if it has 3 points or less
    if (newPolygonCoordinates.length < 4 && isNewPathClosed) {
        newPolygonCoordinates.shift();
    }

    // Remove closing points if either of them was removed
    if (isOldPathClosed) {
        if (selectedPoints.has(0) && !selectedPoints.has(polygonCoordinates.length - 1)) {
            newPolygonCoordinates.pop();
        }
        if (!selectedPoints.has(0) && selectedPoints.has(polygonCoordinates.length - 1)) {
            newPolygonCoordinates.shift();
        }
    }

    // Close previously closed path if it has 3 points or more
    if (isOldPathClosed && !isNewPathClosed && newPolygonCoordinates.length > 2) {
        newPolygonCoordinates.push({ ...newPolygonCoordinates[0] });
    }

    return newPolygonCoordinates;
};

export const getCenterCoordinate = (coordA: Coordinate, coordB: Coordinate): Coordinate => ({
    latitude: (coordA.latitude + coordB.latitude) / 2,
    longitude: (coordA.longitude + coordB.longitude) / 2,
});

// Returns the center coordinates of the polygon edges
export const getPolygonEdges = (polygon: Coordinate[]) =>
    polygon.reduce<Coordinate[]>((edges, coordinate, index) => {
        if (index === 0 || isEqual(polygon[index], polygon[index - 1])) {
            return edges;
        }
        edges.push(getCenterCoordinate(polygon[index], polygon[index - 1]));
        return edges;
    }, []);

// Always returns a list of polygons from a single or multiple polygons
export const ensurePolygonList = (polygons: Coordinate[] | Coordinate[][]): Coordinate[][] => {
    if (polygons.length === 0) {
        return [[]];
    }

    if (isPolygonList(polygons)) {
        // we have to cast here because ts can not infer the type from Array.isArray
        return polygons;
    }

    // we have to cast here because ts can not infer the type from Array.isArray
    return [polygons];
};

export const isPolygonList = (polygons: Coordinate[] | Coordinate[][]): polygons is Coordinate[][] => {
    return Array.isArray(polygons[0]);
};
