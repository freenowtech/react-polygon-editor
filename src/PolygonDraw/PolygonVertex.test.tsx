import { shallow, ShallowWrapper } from 'enzyme';
import { LeafletMouseEvent } from 'leaflet';
import React from 'react';
import { CircleMarker } from 'react-leaflet';

import { Coordinate } from '../types';

import { createLeafletLatLngFromCoordinate } from '../helpers';
import { PolygonVertex, Props, State } from './PolygonVertex';

describe('PolygonVertex', () => {
    const coordinate: Coordinate = { longitude: 0, latitude: 0 };
    const leafletMouseEvent = { latlng: createLeafletLatLngFromCoordinate(coordinate) } as LeafletMouseEvent;
    const props: Props = {
        coordinate: { longitude: 0, latitude: 0 },
        isSelected: true,
        index: 0,

        onClick: jest.fn(),
        onDragStart: jest.fn(),
        onDrag: jest.fn(),
        onDragEnd: jest.fn()
    };

    let wrapper: ShallowWrapper<Props, State, PolygonVertex>;

    beforeEach(() => {
        wrapper = shallow(<PolygonVertex {...props} />);
    });

    describe('When point is NOT selected', () => {
        it('should render correctly', () => {
            wrapper.setProps({ isSelected: false });
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('When point is selected', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('When point is hovered', () => {
        it('should render correctly', () => {
            wrapper.setState({ isHovered: true });
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('should tirgger the onClick callback when user clicks the vertex', () => {
        wrapper.find(CircleMarker).prop('onClick')(leafletMouseEvent);

        expect(props.onClick).toHaveBeenCalledWith(props.index);
    });
});
