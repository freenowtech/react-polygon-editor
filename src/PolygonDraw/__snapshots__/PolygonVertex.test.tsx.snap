// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolygonVertex When point is NOT selected should render correctly 1`] = `
<ForwardRef(Leaflet(CircleMarker))
  bubblingMouseEvents={false}
  center={
    Object {
      "lat": 0,
      "lng": 0,
    }
  }
  color="#096bdb"
  draggable={true}
  fillColor="#ffffff"
  fillOpacity={1}
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  opacity={1}
  radius={4}
  weight={1}
/>
`;

exports[`PolygonVertex When point is hovered should render correctly 1`] = `
<ForwardRef(Leaflet(CircleMarker))
  bubblingMouseEvents={false}
  center={
    Object {
      "lat": 0,
      "lng": 0,
    }
  }
  color="#096bdb"
  draggable={true}
  fillColor="#ffffff"
  fillOpacity={1}
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  opacity={1}
  radius={6}
  weight={4}
/>
`;

exports[`PolygonVertex When point is selected should render correctly 1`] = `
<ForwardRef(Leaflet(CircleMarker))
  bubblingMouseEvents={false}
  center={
    Object {
      "lat": 0,
      "lng": 0,
    }
  }
  color="#096bdb"
  draggable={true}
  fillColor="#ffffff"
  fillOpacity={1}
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  opacity={1}
  radius={6}
  weight={4}
/>
`;
